File: packages/python/holoconf-aws/src/holoconf_aws/ssm.py
"""SSM Parameter Store resolver for holoconf.

Provides the `ssm` resolver for fetching values from AWS Systems Manager Parameter Store.

## Usage

```yaml
database:
    host: ${ssm:/app/prod/db-host}
    password: ${ssm:/app/prod/db-password}
```

## Parameter Types

- **String**: Returned as-is
- **SecureString**: Automatically marked as sensitive for redaction
- **StringList**: Returned as an array (split by comma)

## Kwargs

- `region`: Override the AWS region for this parameter
- `profile`: Use a specific AWS profile
- `default`: Value to use if parameter not found (framework-handled)
- `sensitive`: Override automatic sensitivity detection (framework-handled)
"""

from typing import Any

import boto3
from botocore.exceptions import ClientError

import holoconf

_registered = False


class SsmResolver:
    """SSM Parameter Store resolver.

    Fetches values from AWS Systems Manager Parameter Store.

    Attributes:
        name: The resolver name ("ssm")
    """

    name = "ssm"

    def __init__(self, session: boto3.Session | None = None):
        """Create a new SSM resolver.

        Args:
            session: Optional boto3 session to use. If not provided, a default
                    session will be created.
        """
        self._session = session

    def __call__(self, path: str, **kwargs: Any) -> Any:
        """Resolve an SSM parameter.

        Args:
            path: The SSM parameter path (must start with /)
            **kwargs: Optional resolver kwargs:
                - region: AWS region override
                - profile: AWS profile to use

        Returns:
            The parameter value:
            - String: returned as-is
            - SecureString: returned as ResolvedValue with sensitive=True
            - StringList: returned as a list

        Raises:
            KeyError: If the parameter is not found (triggers default handling)
            ValueError: If the path doesn't start with /
        """
        # Validate path format
        if not path.startswith("/"):
            raise ValueError(f"SSM parameter path must start with /: {path}")

        # Build session with optional overrides
        session = self._get_session(kwargs.get("region"), kwargs.get("profile"))
        client = session.client("ssm")

        try:
            response = client.get_parameter(Name=path, WithDecryption=True)
        except ClientError as e:
            error_code = e.response.get("Error", {}).get("Code", "")
            if error_code == "ParameterNotFound":
                # Raise KeyError to trigger default handling at framework level
                raise KeyError(f"SSM parameter not found: {path}") from e
            raise

        parameter = response.get("Parameter", {})
        value = parameter.get("Value", "")
        param_type = parameter.get("Type", "String")

        # Handle different parameter types
        if param_type == "SecureString":
            # SecureString is automatically sensitive
            return holoconf.ResolvedValue(value, sensitive=True)
        elif param_type == "StringList":
            # StringList is comma-separated, convert to array
            return value.split(",")
        else:
            # Regular string
            return value

    def _get_session(
        self, region: str | None = None, profile: str | None = None
    ) -> boto3.Session:
        """Get a boto3 session with optional overrides."""
        if self._session is not None and region is None and profile is None:
            return self._session

        session_kwargs = {}
        if region:
            session_kwargs["region_name"] = region
        if profile:
            session_kwargs["profile_name"] = profile

        return boto3.Session(**session_kwargs) if session_kwargs else boto3.Session()


def register_ssm(force: bool = False) -> None:
    """Register the SSM resolver with holoconf.

    Args:
        force: If True, overwrite any existing 'ssm' resolver.
              If False (default), only register if not already registered.

    This function is idempotent - calling it multiple times with force=False
    will only register the resolver once.

    Note: The 'ssm' resolver is automatically registered when you import
    the holoconf_aws package, so you typically don't need to call this directly.
    """
    global _registered

    if _registered and not force:
        return

    resolver = SsmResolver()
    holoconf.register_resolver("ssm", resolver, force=force)
    _registered = True

File: crates/holoconf-aws/src/ssm.rs
//! SSM Parameter Store resolver
//!
//! Provides the `ssm` resolver for fetching values from AWS Systems Manager Parameter Store.

use std::collections::HashMap;
use std::sync::Arc;

use aws_config::BehaviorVersion;
use aws_sdk_ssm::types::ParameterType;
use holoconf_core::error::{Error, Result};
use holoconf_core::resolver::{register_global, ResolvedValue, Resolver, ResolverContext};
use holoconf_core::Value;
use tokio::runtime::Runtime;

/// SSM Parameter Store resolver.
///
/// Fetches values from AWS Systems Manager Parameter Store.
///
/// ## Usage
///
/// ```yaml
/// database:
///   host: ${ssm:/app/prod/db-host}
///   password: ${ssm:/app/prod/db-password}
/// ```
///
/// ## Parameter Types
///
/// - **String**: Returned as-is
/// - **SecureString**: Automatically marked as sensitive for redaction
/// - **StringList**: Returned as an array (split by comma)
///
/// ## Kwargs
///
/// - `region`: Override the AWS region for this parameter
/// - `profile`: Use a specific AWS profile
/// - `default`: Value to use if parameter not found (framework-handled)
/// - `sensitive`: Override automatic sensitivity detection (framework-handled)
pub struct SsmResolver {
    runtime: Runtime,
}

impl SsmResolver {
    /// Create a new SSM resolver.
    pub fn new() -> Self {
        let runtime = Runtime::new().expect("Failed to create Tokio runtime");
        Self { runtime }
    }

    /// Fetch a parameter from SSM.
    async fn fetch_parameter(
        &self,
        path: &str,
        region: Option<&str>,
        profile: Option<&str>,
    ) -> Result<(String, ParameterType)> {
        // Build AWS config
        let mut config_loader = aws_config::defaults(BehaviorVersion::latest());

        if let Some(region) = region {
            config_loader = config_loader.region(aws_config::Region::new(region.to_string()));
        }

        if let Some(profile) = profile {
            config_loader = config_loader.profile_name(profile);
        }

        let config = config_loader.load().await;
        let client = aws_sdk_ssm::Client::new(&config);

        // Get parameter with decryption
        let response = client
            .get_parameter()
            .name(path)
            .with_decryption(true)
            .send()
            .await
            .map_err(|e| Error::not_found(format!("SSM parameter '{}': {}", path, e), None))?;

        let parameter = response
            .parameter()
            .ok_or_else(|| Error::not_found(format!("SSM parameter '{}'", path), None))?;

        let value = parameter
            .value()
            .ok_or_else(|| {
                Error::not_found(format!("SSM parameter '{}' has no value", path), None)
            })?
            .to_string();

        let param_type = parameter.r#type().cloned().unwrap_or(ParameterType::String);

        Ok((value, param_type))
    }
}

impl Default for SsmResolver {
    fn default() -> Self {
        Self::new()
    }
}

impl Resolver for SsmResolver {
    fn resolve(
        &self,
        args: &[String],
        kwargs: &HashMap<String, String>,
        _ctx: &ResolverContext,
    ) -> Result<ResolvedValue> {
        // Validate args
        if args.is_empty() {
            return Err(Error::resolver_custom(
                "ssm",
                "SSM resolver requires a parameter path argument",
            ));
        }

        let path = &args[0];

        // SSM paths must start with /
        if !path.starts_with('/') {
            return Err(Error::resolver_custom(
                "ssm",
                format!("SSM parameter path must start with /: {}", path),
            ));
        }

        let region = kwargs.get("region").map(|s| s.as_str());
        let profile = kwargs.get("profile").map(|s| s.as_str());

        // Fetch the parameter using the async runtime
        let (value, param_type) = self
            .runtime
            .block_on(self.fetch_parameter(path, region, profile))?;

        // Handle different parameter types
        match param_type {
            ParameterType::SecureString => {
                // SecureString is automatically sensitive
                Ok(ResolvedValue::sensitive(Value::String(value)))
            }
            ParameterType::StringList => {
                // StringList is comma-separated, convert to array
                let values: Vec<Value> = value
                    .split(',')
                    .map(|s| Value::String(s.to_string()))
                    .collect();
                Ok(ResolvedValue::new(Value::Sequence(values)))
            }
            _ => {
                // Regular string
                Ok(ResolvedValue::new(Value::String(value)))
            }
        }
    }

    fn name(&self) -> &str {
        "ssm"
    }
}

/// Register the SSM resolver in the global registry.
pub fn register() {
    let resolver = Arc::new(SsmResolver::new());
    // Use force=true to allow re-registration (e.g., during testing)
    let _ = register_global(resolver, true);
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_ssm_resolver_name() {
        let resolver = SsmResolver::new();
        assert_eq!(resolver.name(), "ssm");
    }

    #[test]
    fn test_ssm_resolver_no_args() {
        let resolver = SsmResolver::new();
        let ctx = ResolverContext::new("test.path");

        let result = resolver.resolve(&[], &HashMap::new(), &ctx);
        assert!(result.is_err());
        assert!(result
            .unwrap_err()
            .to_string()
            .contains("requires a parameter path"));
    }

    #[test]
    fn test_ssm_resolver_path_must_start_with_slash() {
        let resolver = SsmResolver::new();
        let ctx = ResolverContext::new("test.path");
        let args = vec!["invalid-path".to_string()];

        let result = resolver.resolve(&args, &HashMap::new(), &ctx);
        assert!(result.is_err());
        assert!(result
            .unwrap_err()
            .to_string()
            .contains("must start with /"));
    }

    #[test]
    fn test_register_doesnt_panic() {
        // Just verify registration doesn't panic
        register();
    }

    // Note: Integration tests with actual SSM require AWS credentials
    // and should be run separately with appropriate test fixtures
}
